connection_pool/connection_pool.go

```diff
@@ -32,17 +32,31 @@ var (
 	ErrNoHealthyInstance = errors.New("can't find healthy instance in pool")
 )
 
-/*
-Additional options (configurable via ConnectWithOpts):
-
-- CheckTimeout - time interval to check for connection timeout and try to switch connection.
-*/
+// ConnectionHandler provides callbacks for components interested in handling
+// changes of connections in a ConnectionPool.
+type ConnectionHandler interface {
```

Comment:
What do you think about:
```Go
type ConnectionHandler interface {
    // A connection with a role has been discovered, but not added into a pool yet.
    Discovered(conn *Connection, role Role) error
    // A connection has been removed from a pool already.
    Removed(conn *Connection, role Role) error   
}
```
?

So update is Removed() + Discovered(). It may help to hide the implementation details (Open/Close/etc).

Reply:
I want to call it `RoleHandler`, but actually it is not. We want to add/we removed a connection **with** a role into a pool. We don't want to add a role with a connection...

Reply:
Yeah, "Discovered" is definitely better.

Also I think that all that "removed from a pool"/"added to a pool" stuff is confusing. For a user, there is only one pool -- ConnectionPool object. He shouldn't know about internal implementation and three pools on the inside. On role switch, connection is not removed from the ConnectionPool -- only from some internal pools.

Reply:
Thank you for your patience. Then one more try:
```Go
type ConnectionHandler interface {
    // A connection with a role has been detected, but is not yet available to send requests.
    // It allows for a client to initialize the connection before using it in a pool.
    //
    // The client code may cancel adding a connection to the pool. The client
    // need to return an error from the Discovered call for that. In this case
    // the pool will close connection and will try to reopen it later.
    Discovered(conn *Connection, role Role) error
    // A connection with a role has become unavaileble to send requests. It happens if the
    // connection is closed or the connection role is switched.
    Deactivated(conn *Connection, role Role) error   
}
```
Whould you like this?

Reply:
Yeah, thank you! Maybe it's also worth to add some comment about "role switch triggers Deactivated + Discovered", like it was before.

Reply:
Btw, maybe something like `RoleOn`/`RoleOff` would be nice naming too (up to you to decide)

Reply:
> Btw, maybe something like `RoleOn`/`RoleOff` would be nice naming too (up to you to decide)

It seems to me that longer and more complex names in this case are not too bad idea. It highlights the difference in events and provokes an user to read the documentation for details.

`RoleOn` can be confusing here becase it `BeforeRoleOn` or even `BeforeConnectionOn`.

I've updated the naming and commnets. You could review the changes.

Reply:
Wait, it's Opened/Closed again

Reply:
I somehow pushed the wrong commit... I'm sorry. Updated.